File: go/connector/gdrive_helpers.go
// SPDX-License-Identifier: Apache-2.0

package connector

import (
	"encoding/json"
	"fmt"
	"math"
	"math/rand/v2"
	"net/http"
	"regexp"
	"strconv"
	"time"
)

// folderIDPattern matches valid Google Drive folder IDs (alphanumeric, hyphens, underscores).
var folderIDPattern = regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)

// mimeTypePattern matches valid MIME type strings.
var mimeTypePattern = regexp.MustCompile(`^[a-zA-Z0-9][a-zA-Z0-9!#$&\-.^_+]+/[a-zA-Z0-9][a-zA-Z0-9!#$&\-.^_+]+$`)

// isGoogleNativeFormat reports whether the MIME type is a Google-native
// document format that must be exported rather than downloaded.
func isGoogleNativeFormat(mime string) bool {
	switch mime {
	case mimeGoogleDoc, mimeGoogleSheet, mimeGoogleSlides, mimeGoogleDrawing:
		return true
	}
	return false
}

// parseFileSize parses the file size string from the Drive API. Returns
// 0 for empty or unparsable values (Google-native formats omit size).
func parseFileSize(sizeStr string) int64 {
	if sizeStr == "" {
		return 0
	}
	size, err := strconv.ParseInt(sizeStr, 10, 64)
	if err != nil {
		return 0
	}
	return size
}

// buildFileMetadata constructs the metadata map for a Drive file.
func buildFileMetadata(f *driveFile) map[string]string {
	meta := map[string]string{
		"source":    "gdrive",
		"mime_type": f.MIMEType,
		"file_id":   f.ID,
	}
	if len(f.Parents) > 0 {
		meta["parent_id"] = f.Parents[0]
	}
	if f.Size != "" {
		meta["size"] = f.Size
	}
	return meta
}

// truncateBody returns at most 200 bytes of a response body for error
// messages, avoiding excessively long error strings.
func truncateBody(body []byte) string {
	const maxLen = 200
	if len(body) <= maxLen {
		return string(body)
	}
	return string(body[:maxLen]) + "..."
}

// isRateLimitError checks whether the error response body indicates a
// rate-limit error that should be retried.
func isRateLimitError(body []byte, statusCode int) bool {
	if statusCode == http.StatusTooManyRequests {
		return true
	}

	var apiErr struct {
		Error struct {
			Errors []struct {
				Reason string `json:"reason"`
			} `json:"errors"`
		} `json:"error"`
	}
	if err := json.Unmarshal(body, &apiErr); err != nil {
		return false
	}

	rateLimitReasons := map[string]bool{
		"rateLimitExceeded":     true,
		"userRateLimitExceeded": true,
	}
	for _, e := range apiErr.Error.Errors {
		if rateLimitReasons[e.Reason] {
			return true
		}
	}
	return false
}

// calculateGoBackoff computes the backoff duration with exponential
// increase and jitter, respecting the Retry-After header when present.
func calculateGoBackoff(attempt int, retryAfter string) time.Duration {
	if retryAfter != "" {
		seconds, err := strconv.ParseFloat(retryAfter, 64)
		if err == nil && seconds > 0 {
			d := time.Duration(seconds * float64(time.Second))
			if d > rateLimitMaxBackoff {
				return rateLimitMaxBackoff
			}
			return d
		}
	}

	exponential := float64(rateLimitBaseBackoff) * math.Pow(2, float64(attempt))
	jitter := rand.Float64() * float64(rateLimitBaseBackoff)
	backoff := time.Duration(exponential + jitter)

	if backoff > rateLimitMaxBackoff {
		return rateLimitMaxBackoff
	}
	return backoff
}

// bearerTokenPattern matches Bearer tokens in error messages.
var bearerTokenPattern = regexp.MustCompile(`Bearer\s+[A-Za-z0-9\-._~+/]+=*`)

// sanitiseGoError strips access tokens from error messages to prevent
// credential leakage through error wrapping.
func sanitiseGoError(err error) error {
	msg := err.Error()
	sanitised := bearerTokenPattern.ReplaceAllString(msg, "Bearer [REDACTED]")
	if sanitised != msg {
		return fmt.Errorf("%s", sanitised)
	}
	return err
}

File: sdks/ts/memory/src/connector/gdrive-helpers.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Pure helper functions for the Google Drive connector. Extracted to
 * keep the main connector file under the 700-line limit.
 */

import type { GDriveConnectorConfig } from './gdrive.js'

// -- Constants ----------------------------------------------------------------

/** Maximum response body length for error message truncation. */
const MAX_ERROR_BODY_LENGTH = 200

/** Maximum number of retry attempts on rate-limit errors. */
export const RATE_LIMIT_MAX_RETRIES = 5

/** Base backoff duration in milliseconds for rate-limit retries. */
export const RATE_LIMIT_BASE_BACKOFF_MS = 1_000

/** Maximum backoff duration in milliseconds for rate-limit retries. */
export const RATE_LIMIT_MAX_BACKOFF_MS = 60_000

/** Pattern that valid Google Drive folder IDs must match. */
const FOLDER_ID_PATTERN = /^[a-zA-Z0-9_-]+$/

/** Pattern that valid MIME types must match. */
const MIME_TYPE_PATTERN = /^[a-zA-Z0-9][a-zA-Z0-9!#$&\-.^_+]+\/[a-zA-Z0-9][a-zA-Z0-9!#$&\-.^_+]+$/

const MIME_GOOGLE_FOLDER = 'application/vnd.google-apps.folder'

// -- Types --------------------------------------------------------------------

type DriveAPIError = {
  readonly error: {
    readonly message: string
    readonly errors?: readonly { readonly reason: string }[] | undefined
  }
}

type DriveFile = {
  readonly id: string
  readonly name: string
  readonly mimeType: string
  readonly modifiedTime: string
  readonly size?: string | undefined
  readonly parents?: readonly string[] | undefined
  readonly webViewLink?: string | undefined
}

export type RateLimitParseResult = {
  readonly isRateLimit: boolean
  readonly message: string
}

// -- Functions ----------------------------------------------------------------

export const parseFileSize = (sizeStr: string | undefined): number => {
  if (sizeStr === undefined || sizeStr === '') return 0
  const parsed = Number(sizeStr)
  return Number.isNaN(parsed) ? 0 : parsed
}

export const buildFileQuery = (cfg: GDriveConnectorConfig): string => {
  const parts: string[] = []

  parts.push(`mimeType != '${MIME_GOOGLE_FOLDER}'`)

  if (cfg.folderId) {
    validateFolderId(cfg.folderId)
    parts.push(`'${cfg.folderId}' in parents`)
  }

  if (cfg.mimeTypeFilter !== undefined && cfg.mimeTypeFilter.length > 0) {
    const mimeConditions = cfg.mimeTypeFilter.map((mime) => {
      validateMimeType(mime)
      return `mimeType = '${mime}'`
    })
    parts.push(`(${mimeConditions.join(' or ')})`)
  }

  parts.push('trashed = false')

  return parts.join(' and ')
}

/**
 * Validate that a folder ID matches the expected pattern to prevent
 * query injection via the Drive API query string.
 */
const validateFolderId = (folderId: string): void => {
  if (!FOLDER_ID_PATTERN.test(folderId)) {
    throw new Error(`gdrive: invalid folder ID "${folderId}" — must be alphanumeric with hyphens/underscores`)
  }
}

/**
 * Validate that a MIME type matches the expected format to prevent
 * query injection via the Drive API query string.
 */
const validateMimeType = (mimeType: string): void => {
  if (!MIME_TYPE_PATTERN.test(mimeType)) {
    throw new Error(`gdrive: invalid MIME type "${mimeType}"`)
  }
}

export const buildFileMetadata = (file: DriveFile): Readonly<Record<string, string>> => {
  const meta: Record<string, string> = {
    source: 'gdrive',
    mime_type: file.mimeType,
    file_id: file.id,
  }
  if (file.parents !== undefined && file.parents.length > 0) {
    const firstParent = file.parents[0]
    if (firstParent !== undefined) {
      meta['parent_id'] = firstParent
    }
  }
  if (file.size !== undefined && file.size !== '') {
    meta['size'] = file.size
  }
  return meta
}

export const truncateBody = (body: string): string => {
  if (body.length <= MAX_ERROR_BODY_LENGTH) return body
  return body.slice(0, MAX_ERROR_BODY_LENGTH) + '...'
}

/**
 * Parse a rate limit error response body and determine whether it is
 * a retryable rate limit error.
 */
export const parseRateLimitError = (body: string, statusCode: number): RateLimitParseResult => {
  try {
    const apiErr = JSON.parse(body) as DriveAPIError
    const reasons = apiErr.error.errors ?? []

    const rateLimitReasons: ReadonlySet<string> = new Set([
      'rateLimitExceeded',
      'userRateLimitExceeded',
    ])

    for (const entry of reasons) {
      if (rateLimitReasons.has(entry.reason)) {
        return {
          isRateLimit: true,
          message: `gdrive: rate limit exceeded: ${apiErr.error.message}`,
        }
      }
    }

    return {
      isRateLimit: statusCode === 429,
      message: `gdrive: API error (status ${statusCode}): ${apiErr.error.message}`,
    }
  } catch {
    return {
      isRateLimit: statusCode === 429,
      message: `gdrive: rate limit error (status ${statusCode}): ${truncateBody(body)}`,
    }
  }
}

/**
 * Calculate backoff duration with exponential increase and jitter.
 * Respects the Retry-After header when present.
 */
export const calculateBackoff = (attempt: number, retryAfterHeader: string | null | undefined): number => {
  if (retryAfterHeader !== null && retryAfterHeader !== undefined) {
    const retryAfterSeconds = Number(retryAfterHeader)
    if (!Number.isNaN(retryAfterSeconds) && retryAfterSeconds > 0) {
      return Math.min(retryAfterSeconds * 1_000, RATE_LIMIT_MAX_BACKOFF_MS)
    }
  }

  const exponential = RATE_LIMIT_BASE_BACKOFF_MS * Math.pow(2, attempt)
  const jitter = Math.random() * RATE_LIMIT_BASE_BACKOFF_MS
  return Math.min(exponential + jitter, RATE_LIMIT_MAX_BACKOFF_MS)
}

/**
 * Read a response body with a size limit to prevent unbounded memory
 * allocation. Reads incrementally and aborts if the limit is exceeded.
 */
export const readResponseWithLimit = async (resp: Response, maxBytes: number): Promise<string> => {
  const reader = resp.body?.getReader()
  if (reader === undefined) {
    return resp.text()
  }

  const chunks: Uint8Array[] = []
  let totalBytes = 0
  const decoder = new TextDecoder()

  for (;;) {
    const { done, value } = await reader.read()
    if (done) break

    totalBytes += value.byteLength
    if (totalBytes > maxBytes) {
      reader.cancel().catch(() => {})
      throw new Error(
        `gdrive: response body exceeds ${maxBytes} byte limit (received ${totalBytes} bytes so far)`,
      )
    }
    chunks.push(value)
  }

  const combined = new Uint8Array(totalBytes)
  let offset = 0
  for (const chunk of chunks) {
    combined.set(chunk, offset)
    offset += chunk.byteLength
  }

  return decoder.decode(combined)
}

/**
 * Strip access tokens and authorisation headers from error messages
 * to prevent credential leakage through error wrapping.
 */
export const sanitiseError = (err: unknown): Error => {
  if (!(err instanceof Error)) {
    return new Error(`gdrive: unknown error: ${String(err)}`)
  }

  const bearerPattern = /Bearer\s+[A-Za-z0-9\-._~+/]+=*/g
  const sanitised = err.message.replace(bearerPattern, 'Bearer [REDACTED]')

  if (sanitised !== err.message) {
    return new Error(sanitised)
  }
  return err
}

export const sleep = (ms: number): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve, ms))
